src/material/core/datetime/date-selection-model.ts

```diff
@@ -7,135 +7,108 @@
+  abstract add(date: D): void;
   abstract isValid(): boolean;
+  abstract isSame(other: MatDateSelectionModel<D>): boolean;
 }
 
-export interface DateRange<D> {
-  start: D | null;
-  end: D | null;
-}
+
+//export interface DateRange<D> {
+//  start: D | null;
+//  end: D | null;
+//}
 
 /**
  * Concrete implementation of a MatDateSelectionModel that holds a single date.
  */
-export class MatSingleDateSelectionModel<D> extends MatDateSelectionModel<D> {
-  private _date: D | null = null;
-
-  constructor(adapter: DateAdapter<D>, date?: D | null) {
-    super(adapter);
-    this._date = date === undefined ? null : date;
-  }
+@Injectable()
+export class MatSingleDateSelectionModel<D> extends MatDateSelectionModel<D, D> {
 
-  add(date: D | null) {
-    this._date = date;
+  add(date: D) {
+    this.selection = date;
     this._valueChangesSubject.next();
   }
 
-  compareDate(other: MatSingleDateSelectionModel<D>) {
-    const date = this.asDate();
-    const otherDate = other.asDate();
-    if (date != null && otherDate != null) {
-      return this.adapter.compareDate(date, otherDate);
-    }
-    return date === otherDate;
+  isValid(): boolean {
+    return this.selection != null;
   }
 
-  isComplete() { return this._date != null; }
-
   isSame(other: MatDateSelectionModel<D>): boolean {
     return other instanceof MatSingleDateSelectionModel &&
-        this.adapter.sameDate(other.asDate(), this._date);
-  }
-
-  isValid(): boolean {
-    return this._date != null && this.adapter.isDateInstance(this._date) &&
-           this.adapter.isValid(this._date);
-  }
-
-  asDate(): D | null {
-    return this.isValid() ? this._date : null;
-  }
-
-  setDate(date: D | null) {
-    this._date = date;
-    this._valueChangesSubject.next();
+        this.selection === other.selection;
   }
 }
 
 /**
  * Concrete implementation of a MatDateSelectionModel that holds a date range, represented by
  * a start date and an end date.
  */
-export class MatRangeDateSelectionModel<D> extends MatDateSelectionModel<D> {
-  private _start: D | null = null;
-  private _end: D | null = null;
-
-  constructor(adapter: DateAdapter<D>, start?: D | null, end?: D | null) {
-    super(adapter);
-    this._start = start === undefined ? null : start;
-    this._end = end === undefined ? null : end;
-  }
+@Injectable()
+export class MatRangeDateSelectionModel<D> extends
+  MatDateSelectionModel<DateRange<D>, D> {
 
   /**
    * Adds an additional date to the range. If no date is set thus far, it will set it to the
    * beginning. If the beginning is set, it will set it to the end.
    * If add is called on a complete selection, it will empty the selection and set it as the start.
    */
-  add(date: D | null): void {
-    if (this._start == null) {
-      this._start = date;
-    } else if (this._end == null) {
-      this._end = date;
+  add(date: D): void {
+    if (this.selection.start == null) {
+      this.selection.start = date;
```

Comment:
We don't want to modify the current selection, instead create a new one and do `this.selection = newSelection` at the end (just in case anyone is holding references to old selections)